File: java/MatrixProduct.java
/*
 * Title: Matrix Product
 * Author: Sam Gavis-Hughson
 * Date: 8/28/2016
 * 
 * Given a matrix, find the path from top left to bottom right with the greatest
 * product by moving only down and right.
 * 
 * Execution: javac MatrixProduct.java && java MatrixProduct
 * 
 * For more details, check out http://www.byte-by-byte.com/matrixproduct/
 */

public class MatrixProduct {
    // Bottom up dynamic programming solution
    public static int matrixProduct(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return 0;
        
        // Create cache of min and max product to a given cell
        int[][] maxCache = new int[matrix.length][matrix[0].length];
        int[][] minCache = new int[matrix.length][matrix[0].length];
        
        // Fill caches. We start at the top  left and iteratively find the greatest
        // at smallest path to each subsequent cell by considering the greatest and
        // smallest path to the cells above and to the left of the current cell
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                int maxVal = Integer.MIN_VALUE;
                int minVal = Integer.MAX_VALUE;
                
                // If you're in the top left corner, just copy to cache
                if (i == 0 && j == 0) {
                    maxVal = matrix[i][j];
                    minVal = matrix[i][j];
                }
                
                // If we're not at the top, consider the value above
                if (i > 0) {
                    int tempMax = Math.max(matrix[i][j] * maxCache[i-1][j], matrix[i][j] * minCache[i-1][j]);
                    maxVal = Math.max(tempMax, maxVal);
                    int tempMin = Math.min(matrix[i][j] * maxCache[i-1][j], matrix[i][j] * minCache[i-1][j]);
                    minVal = Math.min(tempMin, minVal);
                }
                
                // If we're not on the left, consider the value to the left
                if (j > 0) {
                    int tempMax = Math.max(matrix[i][j] * maxCache[i][j-1], matrix[i][j] * minCache[i][j-1]);
                    maxVal = Math.max(tempMax, maxVal);
                    int tempMin = Math.min(matrix[i][j] * maxCache[i][j-1], matrix[i][j] * minCache[i][j-1]);
                    minVal = Math.min(tempMin, minVal);
                }
                maxCache[i][j] = maxVal;
                minCache[i][j] = minVal;
            }
        }
        
        // Return the max value at the bottom right
        return maxCache[maxCache.length - 1][maxCache[0].length - 1];
    }
    
    // Sample test cases
    public static void main(String[] args) {
        assert matrixProduct(new int[][]{{1,2,3}, {4, 5, 6}, {7, 8, 9}}) == 2016;
        assert matrixProduct(new int[][]{{-1,2,3}, {4, 5, 6}, {7, 8, 9}}) == -324;
        assert matrixProduct(new int[][]{{-1,2,3}, {4, 5, -6}, {7, 8, 9}}) == 1080;
        
        System.out.println("Passed all tests");
    }
}

File: python/MatrixProduct.py
"""
 Title: Matrix Product
 
Given a matrix, find the path from top left to bottom right with the greatest
product by moving only down and right.
 
Execution: python MatrixProduct.py

For more details, check out http://www.byte-by-byte.com/matrixproduct/

"""
import unittest

def matrix_product(mat):
    """
    Bottom-up dynamic programming solution.
    """
    if len(mat) == 0 or len(mat[0]) == 0:
        return 0

    # Create cache of min and max product to a given cell.
    max_cache = [[1] * len(mat[0])] * len(mat)
    min_cache = [[1] * len(mat[0])] * len(mat)

    # Fill caches. We start at the top left and iteratively find the greatest
    # at smallest path to each subsequent cell by considering the greatest and
    # smallest path to the cells above and to the left of the current cell.
    for i in range(len(mat)):
        for j in range(len(mat[0])):
            max_val = float("inf")
            min_val = float("-inf")

            # If you're in the top left corner, just copy to cache
            if i == 0 and j == 0:
                max_val = mat[i][j]
                min_val = mat[i][j]

            # If we're not at the top, consider the value above.
            if i > 0:
                temp_max = max(mat[i][j] * max_cache[i-1][j], mat[i][j] * min_cache[i-1][j])
                max_val = max(temp_max, max_val)
                temp_min = min(mat[i][j] * max_cache[i-1][j], mat[i][j] * min_cache[i-1][j])
                min_val = min(temp_min, min_val)

            # If we're not on the left, consider the value to the left
            if j > 0:
                temp_max = max(mat[i][j] * max_cache[i][j-1], mat[i][j] * min_cache[i][j-1])
                max_val = max(temp_max, max_val)
                temp_min = min(mat[i][j] * max_cache[i][j-1], mat[i][j] * min_cache[i][j-1])
                min_val = min(temp_min, min_val)

            max_cache[i][j] = max_val
            min_cache[i][j] = min_val
    return max_cache[len(max_cache) - 1][len(max_cache[0]) - 1]
        

mat = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
x = matrix_product(mat)
print(x)
